File: client2/src-tauri/tauri-bridge.js
const { tauri } = window;

const tauriAPI = {
    openScaffoldDirectory: (...args) => tauri.promisified({ cmd: 'tauriAPI', operation: 'openScaffoldDirectory', args }),
    getRootPath: (...args) => tauri.promisified({ cmd: 'tauriAPI', operation: 'getRootPath', args }),
    path: {
        join: (...args) => tauri.promisified({ cmd: 'tauriAPI', operation: 'path.join', args }),
        relative: (...args) => tauri.promisified({ cmd: 'tauriAPI', operation: 'path.relative', args }),
        dirname: (...args) => tauri.promisified({ cmd: 'tauriAPI', operation: 'path.dirname', args }),
        getSeparator: (...args) => tauri.promisified({ cmd: 'tauriAPI', operation: 'path.sep', args }),
    },
    fs: {
        exists: (...args) => tauri.promisified({ cmd: 'tauriAPI', operation: 'fs.existsSync', args }),
        mkdir: (...args) => tauri.promisified({ cmd: 'tauriAPI', operation: 'fs.mkdirSync', args }),
        getFiles: (...args) => tauri.promisified({ cmd: 'tauriAPI', operation: 'fs.getFiles', args }),
    },
    child_process: {
        spawn: (...args) => tauri.promisified({ cmd: 'tauriAPI', operation: 'child_process.spawn', args }),
        kill: (...args) => tauri.promisified({ cmd: 'tauriAPI', operation: 'child_process.kill', args }),
        onStdout: (callback) => tauri.listen('child_process.stdout', (event, data) => callback(data)),
        onStderr: (callback) => tauri.listen('child_process.stderr', (event, data) => callback(data)),
        onExit: (callback) => tauri.listen('child_process.exit', (event, data) => callback(data)),
    },
};

window.tauriAPI = tauriAPI;

File: client/src-tauri/src/tauri_bridge.rs
pub const SOURCE: &'static str = r#"
const { __TAURI__ } = window

const invokeArrayResult = (operation, data, ...args) =>
    __TAURI__.invoke('tauri_api', { operation, args, data })

const invokeSingleResult = async (operation, data, ...args) => {
    return (await invokeArrayResult(operation, data, ...args))[0]
}

const listenEvent = (name, callback) => __TAURI__.event.listen(name, (event) => {
    callback(event.payload)
})

const tauriAPI = {
    openScaffoldDirectory: (...args) => invokeSingleResult('openScaffoldDirectory', [], ...args),
    getRootPath: (...args) => invokeSingleResult('getRootPath', [], ...args),
    getJavas: (...args) => invokeArrayResult('getJavas', [], ...args),
    exportMap: (data, ...args) => invokeSingleResult('exportMap', data, ...args),
    path: {
        join: (...args) => invokeSingleResult('path.join', [], ...args),
        relative: (...args) => invokeSingleResult('path.relative', [], ...args),
        dirname: (...args) => invokeSingleResult('path.dirname', [], ...args),
        getSeperator: (...args) => invokeSingleResult('path.sep', [], ...args)
    },
    fs: {
        exists: (...args) => invokeSingleResult('fs.existsSync', [], ...args),
        mkdir: (...args) => invokeSingleResult('fs.mkdirSync', [], ...args),
        getFiles: (...args) => invokeArrayResult('fs.getFiles', [], ...args)
    },
    child_process: {
        // Combine arguments into one array
        spawn: (...args) => invokeSingleResult('child_process.spawn', [], args[0], args[1], ...args[2]),
        kill: (...args) => invokeSingleResult('child_process.kill', [], ...args),
        onStdout: (callback) => listenEvent('child-process-stdout', callback),
        onStderr: (callback) => listenEvent('child-process-stderr', callback),
        onExit: (callback) => listenEvent('child-process-exit', callback)
    }
}

window.tauriAPI = tauriAPI

console.log('Tauri API injected')
"#;
